website/ui/src/app/services/ecommerce-events.service.ts

```diff
@@ -0,0 +1,360 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Injectable} from '@angular/core';
+import {GoogleTagManagerService} from 'angular-google-tag-manager';
+import {js_beautify} from 'js-beautify';
+import {environment} from 'src/environments/environment';
+import {
+  AddToCart,
+  EcommerceEvent,
+  EcommerceEventName,
+  Item,
+  Purchase,
+  ViewCart,
+  ViewItem,
+} from '../models/ecommerce-events';
+import {Basket, Product, ProductVariant, Products} from '../models/products';
+import {ProductsService} from './products.service';
+
+/**
+ * Service for sending ecommerce events to Google Tag Manager.
+ */
+@Injectable({
+  providedIn: 'root',
+})
+export class EcommerceEventsService {
+  events: string[] = [];
+
+  constructor(
+    private gtmService: GoogleTagManagerService,
+    private productService: ProductsService,
+  ) {}
+
+  /**
+   * For the given products return as Items in ecommerce datalayer schema.
+   * @param products the products to format.
+   * @return an array of Items for all of the products.
+   */
+  private getItems(products: Products): Item[] {
+    const items: Item[] = [];
+    for (const product of Object.values(products)) {
+      const productVariant =
+        this.productService.getDefaultProductVariant(product);
+      items.push(this.getItem(product, productVariant));
+    }
+    return items;
+  }
+
+  /**
+   * For the given basket return as Item in ecommerce datalayer schema.
+   * @param basket the basket to get the items from.
+   * @return an array of Items for all of the products in the basket.
+   */
+  private getItemsFromBasket(basket: Basket): Item[] {
+    const items: Item[] = [];
+    for (const basketProduct of Object.values(basket)) {
+      items.push(
+        this.getItem(
+          basketProduct.product,
+          basketProduct.productVariant,
+          basketProduct.quantity,
+        ),
+      );
+    }
+    return items;
+  }
+
+  /**
+   * For the given product return as an Item in ecommerce datalayer schema.
+   * @param product the product to format.
+   * @param productVariant the product variant to format.
+   * @param quantity optionally provide the quantity to use.
+   * @return an ecommerce Item for of the product.
+   */
+  private getItem(
+    product: Product,
+    productVariant: ProductVariant,
+    quantity = 1,
+  ): Item {
+    return {
+      item_id: productVariant.sku,
+      item_name: product.name,
+      price: productVariant.price,
+      quantity,
+      item_variant: productVariant.name,
+    };
+  }
+
+  /**
+   * Output the event to the console.
+   * @param event the event to output
+   */
+  private logEvent(event: EcommerceEvent): void {
+    console.log('Generated ecommerce event:');
```

Comment:
You can also do `console.log('Generated ecommerce event: ', event);`

Reply:
Good callout. Done.